qutip/solver/heom/bofin_fit.py

```diff
@@ -0,0 +1,1095 @@
+        if the correlation function is provided, and a function if
+        an array is provided.
+
+        The array is needed to run the least squares algorithm, while the
+        the function is used to assign a correlation function to the bosonic
+        bath object.
+        """
+        if callable(C):
+            self._t = t
+            self._C_array = C(t)
+            self._C_fun = C
+        else:
+            self._t = t
+            self._C_array = C
+            _C_fun_r = InterpolatedUnivariateSpline(t, np.real(C))
+            _C_fun_i = InterpolatedUnivariateSpline(t, np.imag(C))
+            self._C_fun = lambda t: _C_fun_r(t) + 1j * _C_fun_i(t)
+
+    def _corr_approx(self, t, a, b, c, d=0):
+        r"""
+        This is the form of the correlation function to be used for fitting.
+
+        Parameters
+        ----------
+        t : np.array or float
+            The times at which to evaluates the correlation function.
+        a : list or np.array
+            A list describing the  real part amplitude of the correlation
+            approximation.
+        b : list or np.array
+            A list describing the decay of the correlation approximation.
+        c : list or np.array
+            A list describing the oscillations of the correlation
+            approximation.
+        d:  A list describing the imaginary part amplitude of the correlation
+            approximation, only used if the user selects if the full_ansatz
+            flag from get_fit is True.
+        """
+
+        a = np.array(a)
+        b = np.array(b)
+        c = np.array(c)
+        d = np.array(d)
+        if (d == 0).all():
+            d = np.zeros(a.shape)
+
+        return np.sum(
+            (a[:, None]+1j*d[:, None]) * np.exp(b[:, None] * t[None, :]) *
+            np.exp(1j*c[:, None] * t[None, :]),
+            axis=0,
+        )
+
+    def get_fit(
+        self,
+        Nr=None,
+        Ni=None,
+        final_rmse=2e-5,
+        lower=None,
+        upper=None,
+        sigma=None,
+        guesses=None,
+        full_ansatz=False
+    ):
+        r"""
+        Fit the correlation function with Ni exponential terms
+        for the imaginary part of the correlation function and Nr for the real.
+        If no number of terms is provided, this function determines the number
+        of exponents based on reducing the normalized root mean squared
+        error below a certain threshold.
+
+        Parameters
+        ----------
+        Nr : optional, int
+            Number of exponents to use for the real part.
+            If set to None it is determined automatically.
+        Ni : optional, int
+            Number of exponents terms to use for the imaginary part.
+            If set to None it is found automatically.
+        final_rmse : float
+            Desired normalized root mean squared error. Only used if Ni or Nr
+            are not specified.
+        lower : list
```

Comment:
It's good to have that `full_ansatz` as an option now! Maybe this should just be a list of length 3 if `full_ansatz` is false?